File: clients/typescript/src/canonical.ts
// canonical.ts reproduces Interlock's ONE canonicalization authority
// (ir.Canonical / ir.writeCanonical in Go) byte-for-byte, so a policy authored
// or transported in TypeScript hashes identically to the Go implementation and
// to the frozen golden corpus. This is deliberately a re-implementation of a
// *deterministic* function — NOT of enforcement. There is no Decide and no
// broker here, and there must never be: enforcement stays the trusted Go
// executable (see the "no foreign enforcement" guardrail).
//
// The scheme (must match ir/ir.go exactly):
//   - object keys sorted by UTF-8 byte order (Go sort.Strings), recursively;
//   - no insignificant whitespace;
//   - each string/key escaped like Go's encoding/json default (HTML-escape ON):
//     " \\ \n \r \t as short escapes; < > & and other control chars < 0x20 and
//     U+2028 / U+2029 as \uXXXX (lowercase hex); everything else verbatim UTF-8;
//   - a trailing newline;
//   - hash = "sha256:" + hex(sha256(canonicalBytes)).

import { createHash } from "node:crypto";

const encoder = new TextEncoder();

// canonicalBytes renders a parsed JSON value (object / array / string / number /
// boolean / null) to Interlock canonical bytes. Input is plain JSON data, never
// a class instance — the generated DTOs are structural and parse straight to it.
export function canonicalBytes(value: unknown): Uint8Array {
  return encoder.encode(serialize(value) + "\n");
}

// hash returns the canonical identity of a value: "sha256:" + hex digest of its
// canonical bytes, matching ir.Policy.Hash / ir.HashBytes tagging.
export function hash(value: unknown): string {
  const digest = createHash("sha256").update(canonicalBytes(value)).digest("hex");
  return "sha256:" + digest;
}

function serialize(v: unknown): string {
  if (v === null || v === undefined) return "null";
  switch (typeof v) {
    case "boolean":
      return v ? "true" : "false";
    case "number":
      return encodeNumber(v);
    case "string":
      return encodeString(v);
    case "object":
      if (Array.isArray(v)) {
        return "[" + v.map(serialize).join(",") + "]";
      }
      return encodeObject(v as Record<string, unknown>);
    default:
      throw new Error(`interlock/canonical: unsupported value type ${typeof v}`);
  }
}

function encodeObject(obj: Record<string, unknown>): string {
  const keys = Object.keys(obj).sort(compareUtf8);
  let out = "{";
  for (let i = 0; i < keys.length; i++) {
    if (i > 0) out += ",";
    out += encodeString(keys[i]) + ":" + serialize(obj[keys[i]]);
  }
  return out + "}";
}

// compareUtf8 orders two strings by their UTF-8 byte sequences, matching Go's
// sort.Strings (which compares bytes). For ASCII keys — all the corpus uses —
// this is identical to a code-unit compare, but we do the real thing so keys
// with multibyte characters still sort exactly as Go would.
function compareUtf8(a: string, b: string): number {
  const ab = encoder.encode(a);
  const bb = encoder.encode(b);
  const n = Math.min(ab.length, bb.length);
  for (let i = 0; i < n; i++) {
    if (ab[i] !== bb[i]) return ab[i] - bb[i];
  }
  return ab.length - bb.length;
}

// encodeNumber preserves integer literals; Interlock's corpus contains no
// numbers, but a Receipt.sequence transported through these types stays exact.
// (Go preserves the JSON number literal via json.Number; JSON.parse has already
// collapsed it to a double by the time we see it, so integers round-trip and
// non-integers are emitted with JS's shortest round-trip form. This branch is
// not exercised by the parity corpus.)
function encodeNumber(n: number): string {
  if (!Number.isFinite(n)) {
    throw new Error("interlock/canonical: non-finite number is not valid JSON");
  }
  return String(n);
}

const HEX = "0123456789abcdef";

// encodeString matches Go encoding/json's default string encoder with HTML
// escaping enabled (the mode ir.go's json.Marshal uses).
function encodeString(s: string): string {
  let out = '"';
  for (const ch of s) {
    const code = ch.codePointAt(0)!;
    switch (ch) {
      case '"':
        out += '\\"';
        continue;
      case "\\":
        out += "\\\\";
        continue;
      case "\n":
        out += "\\n";
        continue;
      case "\r":
        out += "\\r";
        continue;
      case "\t":
        out += "\\t";
        continue;
      case "<":
        out += "\\u003c";
        continue;
      case ">":
        out += "\\u003e";
        continue;
      case "&":
        out += "\\u0026";
        continue;
    }
    if (code < 0x20) {
      out += "\\u00" + HEX[(code >> 4) & 0xf] + HEX[code & 0xf];
      continue;
    }
    // U+2028 LINE SEPARATOR / U+2029 PARAGRAPH SEPARATOR: Go escapes these so the
    // output is safe to embed in JavaScript. Matched by code point to avoid
    // embedding the raw characters in this source file.
    if (code === 0x2028 || code === 0x2029) {
      out += "\\u20" + HEX[(code >> 4) & 0xf] + HEX[code & 0xf];
      continue;
    }
    out += ch;
  }
  return out + '"';
}

File: clients/python/src/interlock_protocol/canonical.py
# canonical.py reproduces Interlock's ONE canonicalization authority
# (ir.Canonical / ir.writeCanonical in Go) byte-for-byte, so a policy authored
# or transported in Python hashes identically to the Go implementation and to the
# frozen golden corpus. This is deliberately a re-implementation of a
# *deterministic* function — NOT of enforcement. There is no decide and no broker
# here, and there must never be: enforcement stays the trusted Go executable
# (see the "no foreign enforcement" guardrail).
#
# The scheme (must match ir/ir.go exactly):
#   - object keys sorted by UTF-8 byte order (Go sort.Strings), recursively;
#   - no insignificant whitespace;
#   - each string/key escaped like Go's encoding/json default (HTML-escape ON):
#     " \\ \n \r \t as short escapes; < > & and other control chars < 0x20 and
#     U+2028 / U+2029 as \uXXXX (lowercase hex); everything else verbatim UTF-8;
#   - a trailing newline;
#   - hash = "sha256:" + hex(sha256(canonical_bytes)).
#
# Stdlib only: hashlib for the digest, and the json module solely to PARSE inputs
# at call sites — the serializer below is hand-written because json.dumps does not
# reproduce Go's HTML-escaping or byte-order key sort.
from __future__ import annotations

import hashlib

__all__ = ["canonical_bytes", "hash"]


def canonical_bytes(value: object) -> bytes:
    """Render a parsed JSON value (dict / list / str / int / float / bool / None)
    to Interlock canonical bytes. Input is plain JSON data, never a class
    instance — the generated TypedDicts are structural and parse straight to it.
    """
    return (_serialize(value) + "\n").encode("utf-8")


def hash(value: object) -> str:
    """Return the canonical identity of a value: "sha256:" + hex digest of its
    canonical bytes, matching ir.Policy.Hash / ir.HashBytes tagging.
    """
    digest = hashlib.sha256(canonical_bytes(value)).hexdigest()
    return "sha256:" + digest


def _serialize(v: object) -> str:
    if v is None:
        return "null"
    # bool is a subclass of int in Python — must be checked BEFORE int.
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, str):
        return _encode_string(v)
    if isinstance(v, int):
        return str(v)
    if isinstance(v, float):
        return _encode_float(v)
    if isinstance(v, list):
        return "[" + ",".join(_serialize(x) for x in v) + "]"
    if isinstance(v, dict):
        return _encode_object(v)
    raise TypeError(f"interlock/canonical: unsupported value type {type(v).__name__}")


def _encode_object(obj: dict) -> str:
    # Sort keys by their UTF-8 byte sequences, matching Go's sort.Strings (which
    # compares bytes). Python compares bytes objects lexicographically by byte
    # value, so this reproduces Go exactly even for multibyte keys. For ASCII
    # keys — all the corpus uses — it coincides with an ordinary string sort.
    keys = sorted(obj.keys(), key=lambda k: k.encode("utf-8"))
    return "{" + ",".join(_encode_string(k) + ":" + _serialize(obj[k]) for k in keys) + "}"


def _encode_float(n: float) -> str:
    # Interlock's corpus contains no numbers; this branch is not exercised by the
    # parity corpus. Reject non-finite values (not valid JSON) and emit the
    # shortest round-trip form otherwise.
    if n != n or n in (float("inf"), float("-inf")):
        raise ValueError("interlock/canonical: non-finite number is not valid JSON")
    return repr(n)


_HEX = "0123456789abcdef"

# Short escapes matching Go encoding/json's default string encoder.
_SHORT = {
    '"': '\\"',
    "\\": "\\\\",
    "\n": "\\n",
    "\r": "\\r",
    "\t": "\\t",
    "<": "\\u003c",
    ">": "\\u003e",
    "&": "\\u0026",
}


def _encode_string(s: str) -> str:
    out = ['"']
    for ch in s:
        short = _SHORT.get(ch)
        if short is not None:
            out.append(short)
            continue
        code = ord(ch)
        if code < 0x20:
            out.append("\\u00" + _HEX[(code >> 4) & 0xF] + _HEX[code & 0xF])
            continue
        # U+2028 LINE SEPARATOR / U+2029 PARAGRAPH SEPARATOR: Go escapes these so
        # the output is safe to embed in JavaScript.
        if code == 0x2028 or code == 0x2029:
            out.append("\\u20" + _HEX[(code >> 4) & 0xF] + _HEX[code & 0xF])
            continue
        out.append(ch)
    out.append('"')
    return "".join(out)
